[#102924550] Normalise provisioning scripts between GCE and AWS

actionjack opened the pull request:
[Normalise provisioning scripts between GCE and AWS](https://www.pivotaltracker.com/story/show/102924550)

**What**

We currently have some discrepancies between GCE and AWS provisioning scripts. This story is to make these as consistent as possible. The pattern we intend to use is the current AWS pattern.

**How this PR should be reviewed**

This PR has been written to be reviewed in the following context:
- I want to move the setting of file permissions out of the provisioner script and into terraform and deprecate the use of `ssh-agent`, because it makes sense to have:
  -  The creation of a file, it's ownership and mode handled in the same location the [file provisioner](https://terraform.io/docs/provisioners/file.html) does not support the change of file permissions by [design and never will](https://github.com/hashicorp/terraform/issues/492) so we have used a [remote-exec](https://terraform.io/docs/provisioners/remote-exec.html).
  -  `ssh-agent` is only needed if we are not using the default location of the `id_rsa` private key.
- I want to create a new shared `provision.sh` script
- I want to install some base packages
- I do not want to redeploy bosh if it is already available
- I want to add details for the cf-release for the spiff templates and to optionally
  upload the release from local.
- I want to upload stemcells and releases
- I want to compile and deploy CF manifest using provision.sh
- Look Ma it's magic! I want to convert the yaml file `terraform_outputs*.yml` to a set of
  variables `terraform_output_*`, so we can use them in the provision script.
- I want to copy the logic from the original `gce/provision.sh` to setup the
  static IP in the BOSH command, only on GCE.
- I want to copy the scripts directory to bastion host
- I want to easily parse the bosh output to list releases and stemcells
- I want to fix the file paths and return codes in commands and lower the timeouts for faster feedback
- I want to restore the copying of account.json when deploying the bastion host
- I want to move the delete-route script to the scripts folder
- I want to check if stemcells and/or releases already exist, if so then skip over installing them
- I want to simply the logic in fix IP script
- I want to enable the new provision.sh script for gce and aws
- I want to set the correct path to /tmp/microbosh-info.json
- I want to ensure that the fix ip routing in GCE rule is defined
- I want to set the manifest file to be bosh-manifest.yml for destroy since I changed it earlier
- I need to correct location of manifest and script files since on a clean build these files are no longer present at the previously named locations
- I want to add an ingress rule to allow bastion to access bosh on it's external IP
  address (i.e. `AWS` Elastic IP)
- I want to Removing test for empty .bosh_config file
  If there an error logging into the bosh director an empty hash is inserted into the `.bosh_config` file. So if there is ever a failure you will never been able to log into the
  director unless the `.bosh_config` file is deleted.

**How to test this PR**

This PR can be tested by starting up a new environment in AWS and GCE, testing and then destroying it, followed by checking the relevant IaaS console to ensure all resources have been clean up correctly e.g.

For AWS:

```
make aws DEPLOY_ENV=$MY_ENVIRONMENT
make test-aws DEPLOY_ENV=$MY_ENVIRONMENT
make destroy-aws DEPLOY_ENV=$MY_ENVIRONMENT
```

For GCE

```
make gce DEPLOY_ENV=$MY_ENVIRONMENT
make test-gce DEPLOY_ENV=$MY_ENVIRONMENT
make destroy-gce DEPLOY_ENV=$MY_ENVIRONMENT
```

**Who should review this PR**

A Rumbustious Rhinoceros, if a rowdy single horned mammal is unavailable then a suitably fiesty member of the core team can serve as a substitute reviewer, it is optional if they have a horn or not.

Any member of the core team can merge this PR with the exception of @jimconner and @keymon who I paired with :p

keymon committed:
Change id_rsa mode in terraform provisioner

We upload and setup the SSH keys when provisioning the VM.

keymon committed:
Add a new shared script for the provision of CF

keymon committed:
Implement installation of dependencies

keymon committed:
Allow to deploy and login in BOSH

Do not redeploy bosh if it is available. We use nc to check the open port
for fast feedback.

keymon committed:
Add code to clone the cf-release repo

We need the cf-release for the spiff templates and to optionally
upload the release from local.

keymon committed:
Upload stemcells and releases in provision.sh

We don't check if the releases are already there because 
parsing `bosh releases` output is painful. We will do it later.

keymon committed:
Compile and deploy CF manifest in provision.sh

keymon committed:
Read terraform outputs from yaml in provision.sh

We convert the yaml file `terraform_outputs*.yml` to a set of
variables `terraform_output_*`, so we can use them in the provision
script.

keymon committed:
Copy the logic to assign static IP in GCE for bosh

We copy the logic from the original `gce/provision.sh` to setup the
static IP in the BOSH command, only on GCE.

As commented by Michal Tekel <mtekel@gmail.com> in 2eb8e8ac6e7783a1ddf8a68a798fbc64740aeee7:

> Add microbosh internal route handling into the provision script. This route re-routes all the traffic to the microbosh external IP via the internal network. Currently we always delete the route first, as it prevents bosh-init configuring the microbosh, because the bosh-init always creates the new VM to apply any changes. The old route would then route traffic to blackhole (old VM, which just bosh-init deleted) and won't let bosh-init connect to new instance to finish configuration. (...)
>    
> On the microbosh server, we configure PREROUTING and OUTPUT iptables rules. PREROUTING translates bosh external IP to the internal eth0 IP, replicating what google 1:1 NAT does. The OUTPUT rule translates all outgoing traffic to external mBosh IP back to eth0. This avoids traffic attempting to go via google routing, which for this instance doesn't really work - there's a route for the external IP pointing back to mBosh (which works correctly for all other instances, but this one) and at the same time there's likely NAT being activated. The packets sent to external mBosh IP never return back. Thus we rewrite them explicitly to eth0 IP, replicating what GCE 1:1 NAT does for externally incoming connections.

keymon committed:
Copy all the scripts directory to bastion

keymon committed:
Add easy to parse scripts to list release/stemcell

keymon committed:
Fix file paths and return codes in commands

keymon committed:
Copy account.json when deploying the bastion host

keymon committed:
Move delete-route script to scripts/ folder

We reuse the functions in gce-assign-fix-ip.sh

keymon committed:
Improve check of existing stemcells and releases

keymon committed:
Fix bug in assign fix ip script

keymon committed:
Enable the new provision.sh script for gce and aws

keymon committed:
Fix minor bugs in assign fixed ip

keymon committed:
Ensure the fix ip routing in GCE rule is defined

keymon committed:
Fix wrong path in bosh-init when destroying

actionjack committed:
Correct location of manifest and script files

On a clean build these files are no longer present at the previously
named location.

actionjack committed:
Add ingress rule for bastion to access bosh

Add ingress rule to allow bastion to access bosh on it's external IP
address (i.e. `AWS` Elastic IP).

This mirrors the same behaviour as `GCE` that does not allow static
internal IP's (at the date of writing).

actionjack committed:
Removing test for empty .bosh_config file

If there an error logging into the bosh director an empty hash is
inserted into the `.bosh_config` file i.e.:

```
---
{}
```

So if there is ever a failure you will never been able to log into the
director unless the `.bosh_config` file is deleted.

saliceti commented on scripts/provision.sh:
```diff
@@ -0,0 +1,260 @@
+
+  echo "Installing binaries: bosh-init, spiff..."
+  if [ ! -x /usr/local/bin/bosh-init ]; then
+    sudo wget -q https://s3.amazonaws.com/bosh-init-artifacts/bosh-init-0.0.72-linux-amd64 -O /usr/local/bin/bosh-init
```
Can you create a variable for the version number and set it at the top

keymon replied:
I will

saliceti commented on scripts/provision.sh:
```diff
@@ -0,0 +1,260 @@
+  fi
+
+  if [ ! -x /usr/local/bin/spiff ]; then
+    wget -q https://github.com/cloudfoundry-incubator/spiff/releases/download/v1.0.7/spiff_linux_amd64.zip
```
Can you create a variable for the version number and set it at the top

saliceti commented on scripts/provision.sh:
```diff
@@ -0,0 +1,260 @@
+    fi
+  fi
+  if ! bosh_check_and_login; then
+    echo "Failed to contact BOSH node $BOSH_IP:$BOSH_PORT after provisioning"
```
Add `return 1` to make the error explicit for the caller (and in other functions too)

keymon replied:
+1 thx!

saliceti commented on scripts/provision.sh:
```diff
@@ -0,0 +1,260 @@
+  time bash $SCRIPT_DIR/deploy_psql_broker.sh admin fakepassword
+}
+
+install_dependencies
```
Create a chain with `cmd1 && cmd2 && cmd3...` so the script stops processing if an error is encountered.

keymon replied:
There is the option `set -e` set, which will make the script stop on ANY error.

saliceti replied:
OK nice one!

keymon committed:
Use variables to define versions of bosh_cli and spiff

keymon committed:
Return failure if not able to login bosh

saliceti commented on scripts/provision.sh:
```diff
@@ -0,0 +1,268 @@
+#   terraform_output_bosh_ip=104.155.37.66
+#
+# etc...
+eval $(
```
This looks a bit hacky. Can we rewrite it in ruby or python?

keymon replied:
Yes you are right, it is too hacky, and it can break if there is a multiline variable, for instance. I will change that.

keymon replied:
That has been fixed. Thx.

keymon committed:
Use a sourced .sh file for terraform output vars

Parsing the terraform_outputs Yaml file in bash is tricky and error
prone. It can break if the Yaml file has some complexity or a
variable with multiple lines.

Instead, we generate from Makefile a valid shell include file with the
required variables, which can be included by `provision.sh`

keymon commented on scripts/provision.sh:
```diff
@@ -0,0 +1,261 @@
+
+BOSH_ADMIN_USER=${BOSH_ADMIN_USER:-admin}
+BOSH_ADMIN_PASS=${BOSH_ADMIN_USER:-admin}
+BOSH_IP=${BOSH_IP:-$terraform_output_bosh_ip}
```
This will cause issues on AWS, because the variable `bosh_ip` in `aws/outputs.tf` is set to the public IP instead of the private IP. 

We need to change `aws/outputs.tf` to use the private IP instead.

saliceti commented:
The `cf` CLI is installed in both `deploy_psql_broker.sh` and `smoke_test.json.sh`. Each one has the version hard coded.
What about moving this to `provision.sh`?

keymon commented:
Yes, I think so, we can move it.

I will do a commit in 5m.

keymon committed:
Move installation of cf_cli to provision.sh

This way we remove duplicated code in deploy_psql_broker and
smoke_test.sh.

Also move the version of bosh_cli with the other version definitions.
